bdns/dns.go

```diff
@@ -283,29 +321,79 @@ func isPrivateV4(ip net.IP) bool {
 	return false
 }
 
-// LookupHost sends a DNS query to find all A records associated with the
-// provided hostname. This method assumes that the external resolver will chase
-// CNAME/DNAME aliases and return relevant A records.  It will retry requests in
-// the case of temporary network errors. It can return net package,
-// context.Canceled, and context.DeadlineExceeded errors.
-func (dnsResolver *DNSResolverImpl) LookupHost(ctx context.Context, hostname string) ([]net.IP, error) {
-	var addrs []net.IP
-	dnsType := dns.TypeA
-	r, err := dnsResolver.exchangeOne(ctx, hostname, dnsType, dnsResolver.aStats)
-	if err != nil {
-		return addrs, &DNSError{dnsType, hostname, err, -1}
+func isPrivateV6(ip net.IP) bool {
+	for _, net := range privateV6Networks {
+		if net.Contains(ip) {
+			return true
+		}
 	}
-	if r.Rcode != dns.RcodeSuccess {
-		return nil, &DNSError{dnsType, hostname, nil, r.Rcode}
+	return false
+}
+
+// LookupHost sends a DNS query to find all A and AAAA records associated with
+// the provided hostname. This method assumes that the external resolver will
+// chase CNAME/DNAME aliases and return relevant records.  It will retry
+// requests in the case of temporary network errors. It can return net package,
+// context.Canceled, and context.DeadlineExceeded errors, all wrapped in the
+// DNSError type.
+func (dnsResolver *DNSResolverImpl) LookupHost(ctx context.Context, hostname string) ([]net.IP, error) {
+	var recordsA, recordsAAAA []dns.RR
+	var errA, errAAAA error
+	var wg sync.WaitGroup
+
+	wg.Add(2)
+	go func() {
```

Comment:
This and the below goroutine block are almost entirely duplicated, it looks like it could be factored to a function quite easily which would make this section much simpler.